Ruby/Ruby.sublime-syntax

```diff
@@ -117,10 +117,10 @@ contexts:
 
   class:
     # Defining a class method
-    - match: \b(class)\b(?=\s*<)
-      scope: keyword.control.class.ruby
-    - match: '\b(class)\b'
-      scope: meta.class.ruby keyword.control.class.ruby
+    - match: \bclass\b(?=\s*<)
+      scope: storage.type.class.ruby
```

Comment:
If I'm correct and this depicts a class method (idk Ruby), then shouldn't this be `storage.modifier`?

Reply:
I found examples the following example, which says something about to use that notation `class << self` just to avoid repetition of the class name in the method declaration.

Not sure whehter it should be scoped differently in such a situation.

```
class TestClass
  # bad
  def TestClass.some_method
    # body omitted
  end

  # good
  def self.some_other_method
    # body omitted
  end

  # Also possible and convenient when you
  # have to define many class methods.
  class << self
    def first_method
      # body omitted
    end

    def second_method_etc
      # body omitted
    end
  end
end
```

Reply:
Okay, that looks *weird*, to say the least. Looks more like a `keyword.control.context` than a class declaration, though.

Reply:
Yep. But I am unsure about using different scopes for the same keyword.

Reply:
Using different scopes for the same keyword is no problem for the `in` operator in Python, for example. We prioritize semantics over syntax anyway and the alternative scope would still be a `keyword`.

Reply:
I would probably argue the other way. Most users are going to find it weird to have the same keyword scoped differently.

Reply:
That's my concern, too.

Reply:
Are you sure about that? `in` is imo a good example,
because it behaves *entirely* different.

```py
for a in b:
    if 1 in a:
        pass
```

Unless you didn't mean to argue for the general case but about `class` in particular here.

Reply:
_Off Topic: I saw several languages scoping `in` as `keyword.operator`. While `keyword.operator.logical` makes sense I am not sure about the use case in the `for` statement._

In `class << self` the `class` keyword has the same meaning as `with` or `using`. It would then need to be scoped as `keyword.control.[block|context]`.

_Sorry, but I still don't feel comfortable with `keyword.control.[exception|resource]` as discussed in #1228 because the former alternatives seem to be a more general approach, which can be applied to more situations/languages without sematical conflicts (see python or the keywords below)._

Reply:
I generally am pretty happy to scope the same symbol differently in different contexts. A great example of this is `=>` in Scala:

```scala
val f: Int => String = i => i.toString
//         ^^ keyword.operator.arrow
//                       ^^ storage.type.function.arrow.lambda
```

I think it's not at all unprecedented.

Ruby is complicated and weird because it has things that are functions but behave like keywords, things that are keywords that behave like functions, and things that are neither that behave differently in different syntactic contexts. In general, I would say that the solution to the *specific* problems raised here (`class` and `in`) should be to scope them differently in the different contexts, biasing toward the differing semantics rather than toward the identical characters.

Reply:
`keyword.control.[exception|resource]` has been replaced by `keyword.control.context.[exception|resource]` (or, ideally, `keyword.context.[exception|resource]`). I just haven't made a definitive statement on that yet because I was collecting input on a new second-level scope.

Reply:
@djspiewak, you should  add `keyword.declaration` to the second arrow, via https://github.com/sublimehq/Packages/pull/2156#discussion_r339352039.